api/v1_projects.inc

```diff
@@ -576,3 +584,114 @@ function api_v1_projects_holdstates($method, $data, $query_params)
 {
     return Project::get_holdable_states();
 }
+
+// checkout a new page to proofread
+function api_v1_project_checkout($method, $data, $query_params)
+{
+    global $pguser;
+
+    $round = $data[":roundid"];
+    $project = $data[":projectid"];
+    validate_for_proofing($round, $project);
+
+    try {
+        $page_array = get_available_proof_page($project, $round, $pguser, $err);
+    } catch (ProjectPageException $exception) {
+        throw new BadRequest($exception->getMessage());
+    }
+    if (!$page_array) {
+        throw new NotFoundError($err, 1005);
+    }
+    [$pagename, $state] = $page_array;
+    // state will get set to temp
+    $proof_page = new ProofPage($project, $round, $pagename);
+    $proof_page->checkout();
+
+    return $proof_page->render_page_data();
+}
+
+function api_v1_project_page($method, $data, $query_params)
+{
+    global $PAGE_BADNESS_REASONS;
+
+    $round = $data[":roundid"];
+    $project = $data[":projectid"];
+    validate_for_proofing($round, $project);
+
+    $pagename = $data[":pagename"];
+    $proof_page = new ProofPage($project, $round, $pagename);
+
+    if ($method == "POST") {
+        // page could be in saved state
+        $proof_page->validate(true);
+        $proof_page->resume_saved_page();
+        return $proof_page->render_page_data();
+    }
+    // page should be in temp state
+    $proof_page->validate(false);
+    if ($method == "GET") {
+        $page_option = get_enumerated_param($query_params, "pageoption", "current", ["current", "original"]);
+        return $proof_page->render_page_text($page_option);
+    }
+    // must be "PUT"
+    $request_data = api_get_request_body();
+    $new_page_state = $request_data["new_page_state"] ?? null;
+    switch ($new_page_state) {
+    case "saved":
+        $text = $request_data["text"] ?? "";
+        $text_checker = new TextChecker($text, $project);
+        if (!$text_checker->all_good) {
+            return $text_checker->response();
+        }
+        $page_saved = $proof_page->attempt_save_as_done($text);
+        if (!$page_saved) {
+            throw new BadRequest(_("You have exceeded your daily page limit, page saved as in progress"), 1009);
```

Comment:
By design, the API currently only returns non-translated strings and we should continue to do so. Clients are responsible for managing the UI presented to the user -- including translated strings. The reason is that the API may be used by an intermediary that is not the user with the API key. For example, PG is working on a tool that will query the API using an automated account as a proxy to the user. The user's desired language is therefore not the one associated with the API key.

Reply:
ok.

Reply:
If we follow the principle that the API should return only data then in the example case of trying to checkout a page there are several possible outcomes:

- success 200 response
-  predictable failure cases where we could return a code and some data, perhaps with the 422 response
-  abnormal errors with other 4xx responses.

The client would have to be able to make a human-readable message from the code and data. This could be done either by pre-loading it with all possible error messages or by an API request to provide a message for for data the client provides, perhaps by another api channel to a possibly different host.

Reply:
There has been some discussion of this topic such as https://groups.google.com/g/api-craft/c/U6JMr26Ew6k and http://soabits.blogspot.com/2013/05/error-handling-considerations-and-best.html.
If we are to provide a user experience at least as good as previously then we will have to provide enough data with the error such as:
- a simple default message to use if all else fails.
- a code for the client to use to construct a more verbose message
- any dynamic data: an array of the parameters for the sprintf used to construct the message in the present system.

Reply:
Thanks for the research! I see the problem with being able to provide localized strings with parameter replacements. If we didn't have to worry about parameter replacements this would be much easier.

Hmm...

What about if failures returned both a non-localized string (the current `error` field) and a localized string based on the user's API key? (`error_localized`?). This would present a consistent client experience for non-web-users and the ability for clients who want a localized string to get one. Thoughts?

Reply:
We could make an extended exception that contained a code and an array of parameters; the existing code could interpret this so we could implement it without breaking the existing function and the api would return this data. The client would have the information to construct messages in the appropriate language

Reply:
That is true, it would be rather complicated. Your proposal would better.

Reply:
I'm sure I don't understand all the ramifications of what's being discussed, but I've been thinking about this and thought I'd ask what type of scenario we're talking about. The current three main users of the API that I can think of at this point are:
1. DP
2. bfoley's scripts that provide some statistics that DP doesn't
3. the script's rfrank is preparing for PG (won't actually be using the API until next PROD deploy, but he's been building them using TEST).

For 1: Will localization be an issue for us when the new PI is in place. I.e., will the API need to be able to localize strings for the PI, or will that be done elsewhere in the code?

For 2: As far as I know, Brian doesn't do any localization.

For 3: Again, I don't think there's any localization going on, and if there is, it would be at their end.

Consider, also, that (1) at this point, only the French translation is enabled, (2) it is intentionally incomplete in the sense that we made a conscious decision to ignore certain classes of messages (for instance, those only seen by site admins or project facilitators are not localized) and (3) maintenance of the translations is only periodic.

So, what is the problem that we're trying to solve?

Reply:
These are the strings and error messages that will be reported to the user in the new PI -- which include strings in the existing PI that we've already translated to French. So without some sort of localization for these strings we will be regressing existing functionality.

Ray's on the right track here because we need to ensure we have a way for the new PI to have translated strings before we move over to it.

Reply:
> Ray's on the right track here because we need to ensure we have a way for the new PI to have translated strings before we move over to it.

Thanks -- that's what I was trying to understand, I think.